### PR TITLE
compression: add per-route remove_accept_encoding_header 

### PR DESCRIPTION
**Pull request by isker**
Commit Message:

```
compression: add per-route remove_accept_encoding_header 

There's no particular reason that this wasn't implemented until now,
other than nobody wanted it. I want it!

The use case: envoy is handling compression for all routes by default,
but some routes _may_ be able to handle compression better than
envoy (e.g. by serving precompressed assets), or they may not, in which
case envoy should still compress. The existing per-route compression
configuration, which only consists of "entirely disabling compression",
is not acceptable for this use case.
```

Additional Description:
Risk Level: Low.
Testing: Unit tests.
Docs Changes: API proto changes.
Release Notes: Yes, included.
Platform Specific Features: No.
[API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):
- Default value of the new proto field ("unset") means this is a no-op by default.

### CODE REVIEW

**Comment by isker** [comment]
https://dev.azure.com/cncf/envoy/_build/results?buildId=150609&view=logs&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=c8fc762c-55bb-57d4-4cab-8ec527e7a423&l=149 Hmm, I have to figure out how to make this work.  If anyone wants to sweep in with an easy answer, please do 🌞.

**Comment by isker** [comment]
> is there a way to avoid this problem entirely?

Maybe `optional` is a way to do this, but I can't find this being used anywhere in the repo.  I am a protobufs ignoramus so I don't know whether it's a bad idea to use it in general, or.

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (toddmgreer) [inline api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Please add the `optional` keyword to this field definition.

The intent of `ResponseDirectionOverrides` is to make it possible to override only what one wants to override, and in particular to make it possible to enable compressor filter per-route, without otherwise changing anything.

There's a reasonable config that would be broken by this change:
  1. You have the compressor disabled at the top level, with `remove_accept_encoding_header: true`.
  2. You have a per-route override providing an empty `response_direction_config`, which enables the compressor.
  3. Now, with this field newly added, the `remove_accept_encoding_header` setting will be overridden.

The reason you don't see `optional` fields is that the proto3 language lacked that feature when Envoy started, so Envoy had to adopt a more cumbersome and less efficient approach when optional scalar fields were required. #14973 discussed adopting `optional`. Idk what further discussion has happened.

If the maintainers are unable to approve the use of `optional`, then we'll need to use the obsolete WKT wrappers.

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (toddmgreer) [inline source/extensions/filters/http/compressor/compressor_filter.cc:183]: This will need to be guarded by `if(config.overrides().response_direction_config().has_remove_accept_encoding_header())`.

File: support/README.md
Comment (toddmgreer) [inline support/README.md:72]: This appears to be an unrelated change. Idk what the maintainers' preferences are--I'm just pointing it out in case it's unintended.

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (toddmgreer) [inline source/extensions/filters/http/compressor/compressor_filter.cc:203]: Now that it's needed in multiple places, it would be best to make per_route_config a member variable (initialized in decodeHeaders). Having done that, it might be more readable to have compressionEnabled and removeAcceptEncodeHeader directly use config_->responseDirectionConfig() and per_route_config_, than to pass them as args.

**Review by toddmgreer — CHANGES_REQUESTED** [review]

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Cool.  I will go ahead and start using `optional`.  We'll see what everyone thinks.

Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Thanks for the review, by the way.

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (isker) [reply source/extensions/filters/http/compressor/compressor_filter.cc:203]: It's okay to keep the raw pointer around as a member?
https://github.com/envoyproxy/envoy/blob/006a20dbd3789339ee8487e304d8f70b61de5508/source/common/http/utility.h#L631-L632

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Welp.  In `//test/tools/type_whisperer:api_type_db_test`:
```
envoy/extensions/filters/http/compressor/v3/compressor.proto: is a proto3 file that contains optional fields, but code generator protoc-gen-api_proto_plugin hasn't been updated to support optional fields in proto3. Please ask the owner of this code generator to support proto3 optional.
```

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (toddmgreer) [reply source/extensions/filters/http/compressor/compressor_filter.cc:203]: hmm... i guess it might be possible for something to change the route between decodeHeaders and encodeHeaders, and it sounds like that would invalidate the pointer. might be best to leave it as-is (because the current form is definitely safe). if anyone has a better understanding of this, feel free to chime in

File: support/README.md
Comment (KBaichoo) [reply support/README.md:72]: +1 can you add to a seperate PR and have phlax@ (who knows about tooling) review it

File: changelogs/current.yaml
Comment (KBaichoo) [inline changelogs/current.yaml:204]: ResponseDirectionOverrides is a nested message in Compressor. which is giving you the docs error

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (KBaichoo) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: This seems like a question for @envoyproxy/api-shepherds  @htuch  AFAIK we haven't moved forward from https://github.com/envoyproxy/envoy/issues/14973

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (KBaichoo) [reply source/extensions/filters/http/compressor/compressor_filter.cc:203]: I've seen bugs sneak in due to route changes etc between decode and encode invalidating pointers to headers, etc. so I'd lean towards the resolving as we're doing.

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this

/wait

File: source/extensions/filters/http/compressor/compressor_filter.cc
Comment (isker) [reply source/extensions/filters/http/compressor/compressor_filter.cc:203]: I will do nothing then.

File: support/README.md
Comment (isker) [reply support/README.md]: https://github.com/envoyproxy/envoy/pull/29990

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (htuch) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: I would like us to move to optional scalar, but someone needs to take #14973 and run with it. I think more than the open discussion there, there is legit work to be done to answer the open questions and fix tooling like our plugins. So, help wanted. For this review, I would just stick with WKT as elsewhere in the API.

Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Can anyone give me an example of achieving optionals with WKT in the codebase?  Blindly grepping didn't get me anything when I tried a while ago.

Comment (htuch) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: I'm just suggesting `google.protobuf.BoolValue ...`

Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:144]: Thanks.  I know nothing about protobuf or C++.

Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto]: This is now using WKT.

File: test/extensions/filters/http/compressor/compressor_filter_test.cc
Comment (isker) [inline test/extensions/filters/http/compressor/compressor_filter_test.cc:432]: I'm open to feedback on making this test less copy-pastey.  I wasn't able to replicate [this similar situation](https://github.com/envoyproxy/envoy/blob/33e7d80206c099fbff898a3402758757cab49436/test/extensions/filters/http/compressor/compressor_filter_test.cc#L188-L211) because it's relying on the top-level filter flag being a runtime flag, which is not the case with `remove_accept_encoding_header`.

**Comment by isker** [comment]
Not sure how the docs errors in the most recent build are related to my changes.

**Review by toddmgreer — APPROVED** [review]

**Review by KBaichoo — DISMISSED** [review]
Thank you for working on this, this looks good to me pending CI and API review

/retest
/wait

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/compressor/v3/compressor.proto:137]: nit: can you please add a link to the `remove_accept_encoding_header` (it appears multiple times here). Maybe it should be this:
```suggestion
  // If set, overrides :ref:`remove_accept_encoding_header<envoy_v3_api_msg_extensions.filters.http.compressor.v3.Compressor.ResponseDirectionConfig.remove_accept_encoding_header>` on the filter's ``ResponseDirectionConfig``.
```

**Review by adisuissa — COMMENTED** [review]
LGTM on the API, modulo minor comment.

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:137]: Sure.  After being thusly punished many times 🌞, I am pretty sure the suggestion will not pass the formatter because the line is too long, but I will do something along these lines.

Comment (adisuissa) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto:137]: That's less of an issue, it's more if I wrote the link correctly.

File: changelogs/current.yaml
Comment (isker) [reply changelogs/current.yaml]: The problem was actually that you need to use `field` not `msg` to refer to fields.  Should now finally be fixed.

File: api/envoy/extensions/filters/http/compressor/v3/compressor.proto
Comment (isker) [reply api/envoy/extensions/filters/http/compressor/v3/compressor.proto]: We got there eventually.

**Review by KBaichoo — APPROVED** [review]

**Comment by isker** [comment]
Thanks to all reviewers!

**Pull request merged**
